File: modules/appattest/android/src/main/java/expo/modules/appattest/AppAttestModule.kt
package expo.modules.appattest

import com.google.android.play.core.integrity.IntegrityManagerFactory
import com.google.android.play.core.integrity.IntegrityTokenRequest
import expo.modules.kotlin.modules.Module
import expo.modules.kotlin.modules.ModuleDefinition
import expo.modules.kotlin.Promise

class AppAttestModule : Module() {
  override fun definition() = ModuleDefinition {
    Name("AppAttest")

    /**
     * Not available on Android — App Attest is iOS-only.
     */
    AsyncFunction("generateKey") { promise: Promise ->
      promise.reject("NOT_AVAILABLE", "Apple App Attest is not available on Android; use getPlayIntegrityToken instead", null)
    }

    /**
     * Not available on Android — App Attest is iOS-only.
     */
    AsyncFunction("attestKey") { _: String, _: String, promise: Promise ->
      promise.reject("NOT_AVAILABLE", "Apple App Attest is not available on Android; use getPlayIntegrityToken instead", null)
    }

    /**
     * Requests a Play Integrity token.
     *
     * @param nonce base64-encoded nonce (base64(SHA-256(challengeBytes)))
     * @returns Play Integrity token string (send to POST /v1/wallets/register as appAttestation.token)
     */
    AsyncFunction("getPlayIntegrityToken") { nonce: String, promise: Promise ->
      val context = appContext.reactContext
        ?: return@AsyncFunction promise.reject("NO_CONTEXT", "No React context available", null)

      val integrityManager = IntegrityManagerFactory.create(context)
      val request = IntegrityTokenRequest.builder()
        .setNonce(nonce)
        .build()

      integrityManager.requestIntegrityToken(request)
        .addOnSuccessListener { response ->
          promise.resolve(response.token())
        }
        .addOnFailureListener { exception ->
          promise.reject("PLAY_INTEGRITY_FAILED", exception.message ?: "Play Integrity request failed", exception)
        }
    }
  }
}

File: modules/appattest/ios/src/AppAttestModule.swift
import DeviceCheck
import ExpoModulesCore

public class AppAttestModule: Module {
  public func definition() -> ModuleDefinition {
    Name("AppAttest")

    /**
     * Generates a new App Attest key pair.
     *
     * Returns the base64-encoded key identifier on success.
     * Rejects if the device does not support App Attest (e.g. simulator).
     */
    AsyncFunction("generateKey") { (promise: Promise) in
      guard DCAppAttestService.shared.isSupported else {
        promise.reject("NOT_AVAILABLE", "DCAppAttestService is not supported on this device")
        return
      }
      DCAppAttestService.shared.generateKey { keyId, error in
        if let error = error {
          promise.reject("GENERATE_KEY_FAILED", error.localizedDescription)
          return
        }
        guard let keyId = keyId else {
          promise.reject("GENERATE_KEY_FAILED", "generateKey returned nil keyId without an error")
          return
        }
        promise.resolve(keyId)
      }
    }

    /**
     * Attests the key with the server challenge.
     *
     * @param keyId              base64-encoded key ID returned by generateKey
     * @param clientDataHashBase64 base64-encoded SHA-256 of the raw challenge bytes
     * @returns base64-encoded CBOR attestation object (send to POST /v1/wallets/register as appAttestation.attestation)
     */
    AsyncFunction("attestKey") { (keyId: String, clientDataHashBase64: String, promise: Promise) in
      guard DCAppAttestService.shared.isSupported else {
        promise.reject("NOT_AVAILABLE", "DCAppAttestService is not supported on this device")
        return
      }
      guard let hashData = Data(base64Encoded: clientDataHashBase64) else {
        promise.reject("INVALID_ARGUMENT", "clientDataHashBase64 is not valid base64")
        return
      }
      DCAppAttestService.shared.attestKey(keyId, clientDataHash: hashData) { attestation, error in
        if let error = error {
          promise.reject("ATTEST_KEY_FAILED", error.localizedDescription)
          return
        }
        guard let attestation = attestation else {
          promise.reject("ATTEST_KEY_FAILED", "attestKey returned nil attestation without an error")
          return
        }
        promise.resolve(attestation.base64EncodedString())
      }
    }

    /**
     * Not available on iOS — Play Integrity is Android-only.
     */
    AsyncFunction("getPlayIntegrityToken") { (_: String, promise: Promise) in
      promise.reject("NOT_AVAILABLE", "Play Integrity is not available on iOS; use generateKey + attestKey instead")
    }
  }
}
